src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -283,9 +280,9 @@ public static string Concat(string? str0, string? str1, string? str2)
             int totalLength = str0.Length + str1.Length + str2.Length;
 
             string result = FastAllocateString(totalLength);
-            FillStringChecked(result, 0, str0);
-            FillStringChecked(result, str0.Length, str1);
-            FillStringChecked(result, str0.Length + str1.Length, str2);
+            CopyStringContent(result, 0, str0);
```

Comment:
Couldn't this now suffer from buffer overruns if the length sums wrap around into a small positive value?

Reply:
Ah good point, can we wrap totalLength calculation into a checked context or that will be a behavioural change? (at least will be foldable for constant inputs)

Reply:
or we can save it to a long and do a quick check + throw the old exception

Reply:
The span-based overloads use checked. It'd be a breaking change to make this use checked, as it'd be changing the type of exception thrown in the case of overflow, but I think it'd be a reasonable breaking change to accept (and document). Your long accumulation idea is also a good one and would avoid the break.

What exception does FastStringAllocate throw for values that are longer than the max allowed string length? It'd be nice if there wasn't a difference in exception type for different forms of "this summed to be too long".

Today strings can't be longer than int.MaxValue / 2, which might limit which of these overloads could actually overflow back to non-negative, but that limit could be removed in the future and it'd be good not to implicitly depend on it here for safety.

Reply:
Good point, the changed behavior is probably not worth the benefit, right? Also, the "checked" context will add new branches we try to avoid here.

I chnaged the PR to only handle `Concat(string,string)` since that one can't physically overflow to a non-negative number. Also, renamed `FillStringChecked` since IMO this names sounds more like it takes a char and fills a string with it, can rename it back if it sounds OK to you

Reply:
> What exception does FastStringAllocate throw for values that are longer than the max allowed string length?

OOM, e.g.:
```csharp
var method = typeof(string).GetMethod("FastAllocateString", 
    BindingFlags.Static | BindingFlags.NonPublic);
method.Invoke(null, new object[] { -1 });
```

Reply:
Got it. 
https://github.com/dotnet/runtime/blob/7c75cbfcaa69d9995799254c7610adb1a0378679/src/coreclr/vm/gchelpers.cpp#L925

Maybe we should just always throw an OOM for an overflow so that the exception is consistent regardless of how much it overflows by. Right now you'll get an OOM up until a certain point and then you'll get an IndexOutOfRangeException beyond that, which is strange. Similarly for the span overloads... you'll get an OOM up until a point and then an OverflowException beyond that.

Reply:
> Got it.
> 
> https://github.com/dotnet/runtime/blob/7c75cbfcaa69d9995799254c7610adb1a0378679/src/coreclr/vm/gchelpers.cpp#L925
> 
> Maybe we should just always throw an OOM for an overflow so that the exception is consistent regardless of how much it overflows by. Right now you'll get an OOM up until a certain point and then you'll get an IndexOutOfRangeException beyond that, which is strange. Similarly for the span overloads... you'll get an OOM up until a point and then an OverflowException beyond that.

If I change that, do I need to file a breaking-change notification?

Reply:
> I need to file a breaking-change notification?

I do not think it is needed in this case. Code out there cannot reasonably depend on where we 

Ideally, we would throw OutOfMemoryException with a message that says the string is too long. It is what we do in other places. For example, `new byte[Int32.MaxValue]` throws `System.OutOfMemoryException: Array dimensions exceeded supported range.`.

Reply:
> I do not think it is needed in this case

I agree.  Since we throw an OOM for int.MaxValue / 2 through int.MaxValue, code that might generate too long a length already needs to be able to handle OOMs if it cares about handling failures, so this is just making things more consistent while not introducing new failure modes.